### ISSUE TITLE
Feature request: Allow logs to be referenced by container name

### ISSUE THREAD

**Issue by JohannesBe** [amir20/dozzle#471]
Hi there,

I'm a huge fan of using dozzle, since it's so lightweight. 
We use it however in a kind of active CI/CD environment, where containers are redeployed constantly. When referring to the logs of a specific container however, the container id is used.

This makes sharing specific container log links not reliable.

Would it be possible to make an endpoint where containers can be referenced by name instead of shortid?

Thanks in advance,

Johannes

**Comment by amir20** [comment]
Hi @JohannesBe, 

Thanks! 

I think I need to understand your use case a little better. The problem of using the name is also conflicting because then you wouldn't be able to see older logs of killed containers. This could be handy if you are trying to figure out why a specific containers was killed. 

Right now, I have chosen the id, because it made more sense to follow a specific instance of container. Using name only, would require for dozzle to convert name to id. Which means it would have to be smart enough to wait for a container to have stopped and then restart the logs with the new id. 

Let's start with telling me more about your team. How do you guys share links? Are these shared in a chat room or something? Would it make sense to have a button that says share this view where all the open logs could also be shared? 

> Would it be possible to make an endpoint where containers can be referenced by name instead of shortid?

Did you mean keep everything using container ids, but have a special end point that looks up via names? Were you thinking this would be a hidden API or also somehow exposed in the UI?

**Comment by JohannesBe** [comment]
Hi @amir20 ,

Thanks for your quick and open response.

I understand where you are coming from: id's are not the same as names. 

The setup I'm talking about is a docker-compose based set-up, where old containers are automatically removed when not running any more.

The reason for this feature request is that, we have a CI/CD set-up, that automatically deploys merge request code to test containers, such that reviewing developers, as well as product owners can test the merge requests easily. This saves a lot of time, since the developer does not have to pull the codebase first and set-up new migrations / dummy database for the MR, and in some cases the PO can preemptively filter some undesired functionalities.

This is coupled with a bot that automatically comments the testing link to the relevant merge request. To ease the developer experience even more, I use dozzle and provide the developers with a logging endpoint. This means that in over 80% of cases, developers do not even have to download any code, config, test-data, etc and are still able to profoundly review the merge request.

Since there is quite a lot of activity (usually about 10-15 merge requests are opened per day, and ~40 are open at the same time), many merge request containers are visible for the developers. This makes finding the logging endpoint somewhat difficult.

The problem that I am struggling with is that:
 1) the merge request endpoints are fixed (e.g. `server.com/mr/2479?token=xxxxxxxxxxx`)
 2) the bot comments the relevant URL only once, as soon as the merge request is opened  
   => this is usually before the automated builds have finished (e.g. the pipeline is still running and the container does not even exist) 

So for me an ideal scenario would be that I could link instead to some sort of endpoint that detects **running** containers with the provided name. Whether this API is hidden or not is not really of importance, but I presume it's best practice to document it.

Kind regards,

Johannes

**Comment by amir20** [comment]
Interesting. I need to think of this use case. At my last company we had a similar setup so I am all familiar with what you are trying to do here. Sounds like a nice culture BTW. :) 

So that I understand...your CI/CD picks up the pull request, which triggers a bot comment. **BEFORE** the PR is merged, you have something like a "sandbox" environment that show the code live. Is that correct? You bring up using compose and then use Dozzle for engineers to jump right to the logs for debugging. Right?

One solution could just simply provide an API like `/show?name=foo` which would internally search for a container and redirect to appropriate container. I am not sure where in the UI I could show this. Maybe just document it.

**Comment by JohannesBe** [comment]
> BEFORE the PR is merged, you have something like a "sandbox" environment that show the code live.  

Yes, exactly! 
During the lifetime of the MR new code pushes might retrigger the pipeline also, rendering old container id endpoints useless for me.

> One solution could just simply provide an API like /show?name=foo which would internally search for a container and redirect to appropriate container. I am not sure where in the UI I could show this. Maybe just document it.

For my use case this would be great! Sadly I'm no good of a go coder so it'd be hard for me to help you with the implementation.

I think one added advantage of these kinds of endpoints could also be to have some sort of fixed link for certain containers, for example, one might also want to use this kind of url when referring to logs in a wiki page or so, uninfluenced by time or new deployments.

Maybe a little share button with some could be useful?

**Comment by amir20** [comment]
Thanks for the update. 

One more question. How do you deploy Dozzle? Is one single instance used to view the logs for all our your PR deployments? 

I wonder if you have thought about using a dedicated Dozzle instance for each PR deployment? If so, you can use the `--filter` option to narrow down to that deployment. But this depends how your URLs work.

**Comment by JohannesBe** [comment]
Hey Amir,

We use one log environment per merge environment, to keep pipelines as short as possible; our current filter is like this: `name=mr--environment--projectname--*`

If you were to implement this, I thought it could be interesting to do something similar like imgur:

![image](https://user-images.githubusercontent.com/21303311/82705625-4708f780-9c78-11ea-9ccb-7ae057aa4636.png)

So a small share button (e.g. something like this)

![image](https://user-images.githubusercontent.com/21303311/82706754-c5669900-9c7a-11ea-97a2-13837c4d37a4.png)


That opens a separate window, in which both urls are shown:
- A field with the container id based url
- A field with the container name based url

Maybe we could add like an (I) or (?) Icon to explain the difference:

![image](https://user-images.githubusercontent.com/21303311/82706851-170f2380-9c7b-11ea-986b-fabd499ad802.png)
![image](https://user-images.githubusercontent.com/21303311/82707051-88e76d00-9c7b-11ea-9bda-f6f6e9339407.png)

Any opnions on this?

**Comment by JohannesBe** [comment]
By the way, I looked up the icomoon share icon (since you seem to be using these):

```
<symbol id="icon-share" viewBox="0 0 32 32">
<path d="M27 22c-1.411 0-2.685 0.586-3.594 1.526l-13.469-6.734c0.041-0.258 0.063-0.522 0.063-0.791s-0.022-0.534-0.063-0.791l13.469-6.734c0.909 0.94 2.183 1.526 3.594 1.526 2.761 0 5-2.239 5-5s-2.239-5-5-5-5 2.239-5 5c0 0.269 0.022 0.534 0.063 0.791l-13.469 6.734c-0.909-0.94-2.183-1.526-3.594-1.526-2.761 0-5 2.239-5 5s2.239 5 5 5c1.411 0 2.685-0.586 3.594-1.526l13.469 6.734c-0.041 0.258-0.063 0.522-0.063 0.791 0 2.761 2.239 5 5 5s5-2.239 5-5c0-2.761-2.239-5-5-5z"></path>
</symbol>
```

**Comment by JohannesBe** [comment]
Another option could be to show the share icon on a per-container basis in the left sidebar, e.g. when hovering an item from the list.

**Comment by amir20** [comment]
Thanks.

So two thoughts: 

* re: Icons: I am hesitant to add any buttons or icons to the main page if I am not convinced that it will be used by at least 50% of users. I haven't gotten many people asking for this feature so I am not sure if it makes sense to have a share button. I have found most people to just use Dozzle to check logs in real-time and then exit. Adding any icons will take space and probably add to confusion with all things going on. Trying to keep Dozzle UI simple has been a challenge :)

* re: Using one Dozzle per PR. I am wondering why you need this feature then? If you are currently deploying multiple Dozzle instances (ie. one per deployment) then don't you already have a direct link to Dozzle instance with all containers already filtered down? For example, I would imagine each PR deployment would be `pr-123.dev.mycompany.com` and you could server dozzle from `/dozzle` instead. Wouldn't that be simpler? Or do you need to link to a specific container for each deployment from your prs? I believe this is called "deep linking" which would allow you to highlight or select an existing item on the page. 

I imagine if you have dozens of containers per PR then deep linking would be nice. I am leaning towards doing `/show?name=my-awesome-container`. 


Thoughts?

**Comment by JohannesBe** [comment]
On second thought I agree that it's maybe for the better not to add the icon. One of the reasons I immediately fell in love with dozzle is due to it's simplicity.

Currently the project with the most complex container PR structure in our company is one with 6 differenr container types (a queue, rev proxy, backend, db and two microservices), but this is due to increase soon. So indeed also for  direct linking in multi container environments this would be very nice.

So in summary I would love the aforementioned endpoint (`/show?name=my-awesome-container`) for two reasons:
1) docker environments or clusters with many (different) containers become directly and consistently addressable by container name
2) links are sharable and more consistent over time

If I can be of any help for this, let me know. I am willing to do a little deepdive into go in my spare time, because that's been on my knowledge agenda for a long time anyway 😉

Anyway, I'll be off for the night now. Will follow up tomorrow.

Kind regards,

Johannes

**Comment by amir20** [comment]
Alright I have merged #474 which should get you `show?name=test`. 

Docker Hub builds every snapshot. So try `dozzle:latest-dev` and see if that does the trick for you. If it does, I can release an official build tomorrow. 

Regarding one Dozzle per PR, I have kept Dozzle light for that exact reason. I thought it would be cool if you add Dozzle to your compose file and use some filter trickery to only show containers for that compose file. But I guess that's more difficult if some nodes are shared. 

You can use tagging by the way. Not sure if you thought of that.

**Comment by JohannesBe** [comment]
Hi Amir, works like a charm!

I should've thought of a frontend-only solution: I knew `container.json` already contained the necessary information. In that case I could've scrambled together a PR myself. Kinda disappointed that I did not notice that myself ¯\_(ツ)_/¯

One small remark though: you use `.includes(this.$route.query.name)`, is this intentional? This means that all names will be wildcarded by default. In the case of the following example for instance:
`mr--myenv--myproj--1` and `mr--myenv--myproj--11`, this could yield unexpected behaviour in my case (although the chances are very slim). Thoughts?

**Comment by amir20** [comment]
Hello, @JohannesBe. That's great to hear. 

All the routing and pages are handled with Vue. Go is only for the API. :) 

I see what you are saying about `includes`. I thought it would be easier if someone wanted to match a pattern. But I'll stick to `==`. 

Let me fix that. And I will push a new build today.

**Comment by amir20** [comment]
`1.26.0` is now up.

**Comment by JohannesBe** [comment]
Thanks, I'll update my dockerfiles (I think it's best practice to use fixed versions in dockerfiles to guarantee future stability, so I was a bit hesitant to use `latest-dev`)!

P. S. In the README.md are some notes regarding reverse proxy config with nginx, I have some potentially useful information on this topic, do you think it is best to open a new issue for this?

**Comment by amir20** [comment]
Great. 

Yea, I wouldn't use `latest-dev` in any compose files. It's just `master` and too much can break. 

Sure, open another bug. I recently discovered `X-Accel-Buffering: no` so I think most issues with Nginx should be fixed since `1.23`. But if there is something new you have found, I'd love to hear. 

I am closing this bug.

**Comment by amir20** [comment]
Ping @JohannesBe, just checking on the issue you mentioned. Did you still want to create the issue?

**Comment by JohannesBe** [comment]
Hi @amir20 , I will create it soon, sorry for keeping you waiting :) Currently quite busy. Will probably be something for the weekend.
Thanks for the ping!

**Comment by JohannesBe** [comment]
Hi @amir20, thanks once again for implementing the feature. It's been working flawlessly for the past month :)
